File: backend/instruments.py
"""Instrument metadata shared across engine, market data, and outcomes."""
from __future__ import annotations

from typing import Literal

Instrument = Literal["XAUUSD", "BTCUSD", "NZDUSD"]

INSTRUMENTS: list[Instrument] = ["XAUUSD", "BTCUSD", "NZDUSD"]

META: dict[str, dict] = {
    "XAUUSD": {
        "label": "Gold",
        "symbol": "XAU/USD",
        "mintick": 0.01,
        "pip_mult": 10.0,
        "decimals": 2,
        "sl_buffer_pips": 20.0,
        "risk_cap_pips": 150.0,
        "dollar_per_pip_per_lot": 10.0,
    },
    "BTCUSD": {
        "label": "Bitcoin",
        "symbol": "BTC/USD",
        "mintick": 1.0,
        "pip_mult": 1.0,
        "decimals": 0,
        "sl_buffer_pips": 0.0,
        "risk_cap_pips": 234.0,
        "dollar_per_pip_per_lot": 0.01,
    },
    "NZDUSD": {
        "label": "NZD/USD",
        "symbol": "NZD/USD",
        "mintick": 0.00001,
        "pip_mult": 10.0,
        "decimals": 5,
        "sl_buffer_pips": 0.0,
        "risk_cap_pips": 80.0,
        "dollar_per_pip_per_lot": 10.0,
    },
}


def pip_size(instrument: str) -> float:
    m = META[instrument]
    return m["mintick"] * m["pip_mult"]


def round_price(instrument: str, price: float) -> float:
    return round(price, META[instrument]["decimals"])


def instrument_label(instrument: str) -> str:
    return META.get(instrument, {}).get("label", instrument)

File: frontend/src/utils/instruments.js
export const INSTRUMENTS = ['XAUUSD', 'BTCUSD', 'NZDUSD']

export const META = {
  XAUUSD: {
    label: 'Gold',
    short: 'GOLD',
    pipSize: 0.1,
    pipLabel: 'pips',
    decimals: 2,
    dollarPerPipPerLot: 10,
    defaultEntry: 2350,
    defaultSl: 2340,
    inputStep: 0.01,
  },
  BTCUSD: {
    label: 'Bitcoin',
    short: 'BTC',
    pipSize: 1,
    pipLabel: 'points',
    decimals: 0,
    dollarPerPipPerLot: 0.01,
    defaultEntry: 105000,
    defaultSl: 104500,
    inputStep: 1,
  },
  NZDUSD: {
    label: 'NZD/USD',
    short: 'NZD',
    pipSize: 0.0001,
    pipLabel: 'pips',
    decimals: 5,
    dollarPerPipPerLot: 10,
    defaultEntry: 0.61250,
    defaultSl: 0.61150,
    inputStep: 0.00001,
  },
}

export function instrumentLabel(id) {
  return META[id]?.label ?? id
}

export function instrumentShort(id) {
  return META[id]?.short ?? id
}

export function pipSize(instrument) {
  return META[instrument]?.pipSize ?? 0.0001
}

export function pipLabel(instrument) {
  return META[instrument]?.pipLabel ?? 'pips'
}

export function formatPrice(instrument, price) {
  if (price == null || Number.isNaN(price)) return '—'
  const dec = META[instrument]?.decimals ?? 2
  return price.toFixed(dec)
}

export function calcDollarRisk(instrument, lotSize, entry, target) {
  const distance = Math.abs(entry - target) / pipSize(instrument)
  return distance * lotSize * (META[instrument]?.dollarPerPipPerLot ?? 10)
}

export function calcRiskDistance(instrument, entry, sl) {
  return Math.abs(entry - sl) / pipSize(instrument)
}

export function mt5Symbol(instrument) {
  return instrument
}
